homeassistant/components/frank_energie/sensor.py

```diff
@@ -0,0 +1,397 @@
+        translation_key="invoice_previous_period",
+        device_class=SensorDeviceClass.MONETARY,
+        state_class=SensorStateClass.TOTAL,
+        native_unit_of_measurement=CURRENCY_EURO,
+        authenticated=True,
+        service_name=SERVICE_NAME_COSTS,
+        value_fn=lambda data: data.invoices.previousPeriodInvoice.TotalAmount
+        if data.invoices.previousPeriodInvoice
+        else None,
+        attr_fn=lambda data: {
+            "Start date": data.invoices.previousPeriodInvoice.StartDate,
+            "Description": data.invoices.previousPeriodInvoice.PeriodDescription,
+        },
+    ),
+    FrankEnergieEntityDescription(
+        key="invoice_current_period",
+        translation_key="invoice_current_period",
+        device_class=SensorDeviceClass.MONETARY,
+        state_class=SensorStateClass.TOTAL,
+        native_unit_of_measurement=CURRENCY_EURO,
+        authenticated=True,
+        service_name=SERVICE_NAME_COSTS,
+        value_fn=lambda data: data.invoices.currentPeriodInvoice.TotalAmount
+        if data.invoices.currentPeriodInvoice
+        else None,
+        attr_fn=lambda data: {
+            "Start date": data.invoices.currentPeriodInvoice.StartDate,
+            "Description": data.invoices.currentPeriodInvoice.PeriodDescription,
+        },
+    ),
+    FrankEnergieEntityDescription(
+        key="invoice_upcoming_period",
+        translation_key="invoice_upcoming_period",
+        device_class=SensorDeviceClass.MONETARY,
+        state_class=SensorStateClass.TOTAL,
+        native_unit_of_measurement=CURRENCY_EURO,
+        authenticated=True,
+        service_name=SERVICE_NAME_COSTS,
+        value_fn=lambda data: data.invoices.upcomingPeriodInvoice.TotalAmount
+        if data.invoices.upcomingPeriodInvoice
+        else None,
+        attr_fn=lambda data: {
+            "Start date": data.invoices.upcomingPeriodInvoice.StartDate,
+            "Description": data.invoices.upcomingPeriodInvoice.PeriodDescription,
+        },
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Frank Energie sensor entries."""
+    frank_coordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    # Add an entity for each sensor type, when authenticated is True,
+    # only add the entity if the user is authenticated
+    async_add_entities(
+        [
+            FrankEnergieSensor(frank_coordinator, description, config_entry)
+            for description in SENSOR_TYPES
+            if not description.authenticated or frank_coordinator.api.is_authenticated
+        ],
+        True,
+    )
+
+
+class FrankEnergieSensor(CoordinatorEntity, SensorEntity):
+    """Representation of a Frank Energie sensor."""
+
+    _attr_attribution = "Data provided by Frank Energie"
+    _attr_has_entity_name = True
+    _attr_icon = "mdi:currency-eur"
+
+    _unsub_update: Callable[[], None] | None = None
```

Comment:
I see that you have a whole update thing going on, why do we do this? The coordinator already updates every hour right?

Reply:
I didn't implement that, but I suspect that it was being added to sync the update to every full hour. @bajansen, any ideas?

Reply:
In the earliest versions of the integration there were no automatic updates but they had to be manually updated through `homeassistant.update_entity`. I suspect it might be related to that and can probably be removed.